### ISSUE TITLE
Inference problems

### ISSUE THREAD

**Issue by bofang68** [wenxi-yue/SurgicalSAM#15]
Hi, I read your inference function. I noticed you first gave the program the specific class before inference (e.g., predicting category 3 for image 1). Is this controversial because it reduces the possibility of misclassification? For example, if image 1 has only class 3 and class 7, it won't produce a predicted mask for class 2 (even though it cannot distinguish between classes 2 and 3, 7). In other words, the authors tell the model which classes the predicted image contains (based on the true labels). Yet most images contain only a few classes.

As far as I'm concerned, it's fair to predict all the classes of an image (e.g., from category 1 to 7) one by one. If it performs well, this model should not predict classes that the image does not contain.

Looking forward to your reply.

![1716479236589](https://github.com/wenxi-yue/SurgicalSAM/assets/170651089/7bd0b025-f753-4221-b9a1-f6374344727c)

**Comment by wenxi-yue** [comment]
Hi, Thanks for your interest. 

We think that these represent two different settings. The setting you mentioned pertains to traditional semantic or instance segmentation, whereas in our paper, we adopt a promptable setting where the user specifies the class of interest. 

Due to the different settings, we divide the methods in our comparison into specialist models (with a traditional setting) and SAM-based models (with a promptable setting). We consistently maintain a promptable setting with all SAM-based models to ensure a fair comparison, and we explicitly clarify this in the paper: "Challenge IoU measures the IoU between the predicted and ground-truth masks for only the classes present in an image, whereas IoU is computed across all classes. In our class promptable segmentation setting with class prompts provided, Challenge IoU and IoU yield identical results."

**Comment by bofang68** [comment]
Thank you for your response. I have an additional problem with the replication process. I followed data_prepocess.py to preprocess the images and labels of the training set and get the format (class_embeddings_h and sam_features_h) consistent with the uploaded dataset. The results for the validation set are poor. Why is this? Should I do the processing for both the validation set and the training set together?
![image](https://github.com/wenxi-yue/SurgicalSAM/assets/170651089/759dd878-3f24-419d-ae2c-f835ffcb5952)

**Comment by bofang68** [comment]
I can understand that you trained through a subset of 40 groups and found that a certain folder had the best training (named it 0) right?

**Comment by wenxi-yue** [comment]
> Thank you for your response. I have an additional problem with the replication process. I followed data_prepocess.py to preprocess the images and labels of the training set and get the format (class_embeddings_h and sam_features_h) consistent with the uploaded dataset. The results for the validation set are poor. Why is this? Should I do the processing for both the validation set and the training set together? ![image](https://private-user-images.githubusercontent.com/170651089/334622727-759dd878-3f24-419d-ae2c-f835ffcb5952.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTc5MDg3ODAsIm5iZiI6MTcxNzkwODQ4MCwicGF0aCI6Ii8xNzA2NTEwODkvMzM0NjIyNzI3LTc1OWRkODc4LTNmMjQtNDE5ZC1hZTJjLWY4MzVmZmNiNTk1Mi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNjA5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDYwOVQwNDQ4MDBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT05NTM5NzRhZDdjM2E2MmFjMmUwMzAzNGZkYThiZWFmOTU3MjhjNDZhNzEzYTdmYTE0ODI1NTgwYTA4MWY3YTdlJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.-KyZKupW3m52cJ73YwPNTpzDRLcnFMuY9cLS3dOo1iA)

For the validation set, you can use the provided validation data directly without any additional processing.

Regarding the poor performance, I may need more details to identify the problem accurately. However, I recommend double-checking that the images and annotations in the training and validation sets are consistent?

**Comment by wenxi-yue** [comment]
> I can understand that you trained through a subset of 40 groups and found that a certain folder had the best training (named it 0) right?

Folder 0 is not the folder with the best performance; it is the original data without any data augmentations. Please read [this comment](https://github.com/wenxi-yue/SurgicalSAM/issues/13#issuecomment-1986705610) for more detailed explanations of what these folders of different versions represent.
